File: aocJulia/days/day03/part2.jl
using Chain, Test

using aocJulia: helpers


function priority(c)
    if isuppercase(c)
        return Int(c) - 38
    else
        return Int(c) - 96
    end
end

function compute(s)
    sacks = s |> chomp |> split .|> collect
    sacks = map.(x -> priority(x), sacks)
    badges = []
    for i in 1:3:length(sacks)
        push!(badges, intersect(sacks[i], sacks[i+1], sacks[i+2]))
    end
    return sum(badges)[1]
end


INPUTS_S = """
vJrwpWtwJgWrhcsFMMfFFhFp
jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL
PmmdzqPrVvPwwTWBwg
wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn
ttgJtRGJQctTZtZT
CrZsJsPPZsGzwwsLwLmpwMDw
"""
EXPECTED = 70

@test compute(INPUTS_S) == EXPECTED

INPUT_TXT = helpers.read_input(dirname(Base.source_path()))
println(compute(INPUT_TXT))

File: aoc-python/day03/part2.py
from __future__ import annotations

import argparse
import os.path

import numpy as np
import pytest

import support

INPUT_TXT = os.path.join(os.path.dirname(__file__), 'input.txt')


def prioritise(c: str) -> int:
    if c.islower():
        return ord(c) - 96
    else:
        return ord(c) - 38


def compute(s: str) -> int:
    lines = s.splitlines()
    in_common = []
    for i in range(0, len(lines), 3):
        elfs = (lines[i], lines[i+1], lines[i+2])
        a, b, c = (set(map(prioritise, e)) for e in elfs)
        in_common.append(a & b & c)
    
    return sum(map(sum, in_common))


INPUT_S = '''\
vJrwpWtwJgWrhcsFMMfFFhFp
jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL
PmmdzqPrVvPwwTWBwg
wMqvLMZHhHMvwLHjbvcjnnSBnvTQFn
ttgJtRGJQctTZtZT
CrZsJsPPZsGzwwsLwLmpwMDw
'''
EXPECTED = 157


@pytest.mark.parametrize(
    ('input_s', 'expected'),
    (
        (INPUT_S, EXPECTED),
    ),
)
def test(input_s: str, expected: int) -> None:
    assert compute(input_s) == expected


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument('data_file', nargs='?', default=INPUT_TXT)
    args = parser.parse_args()

    with open(args.data_file) as f, support.timing():
        print(compute(f.read()))

    return 0


if __name__ == '__main__':
    raise SystemExit(main())
